### PR TITLE
coll: Add Rabenseifner's algorithm for Reduce and Allreduce

### PR DESCRIPTION
**Pull request by mkurnosov**
A component with implementation of R. Rabenseifner's algorithm for Reduce and Allreduce.
This algorithm is a combination of a reduce-scatter implemented with recursive vector halving
and recursive distance doubling, followed either by a gather or an allgather.

Current limitations:
  -- count >= 2^{\floor{\log_2 p}}
  -- commutative operations only
  -- intra-communicators onl

Signed-off-by: Mikhail Kurnosov <mkurnosov@gmail.com>

### CODE REVIEW

**Comment by bwbarrett** [comment]
bot:ompi:retest

**Comment by bwbarrett** [comment]
spacc?  My acronym decoder is failing me...

**Comment by mkurnosov** [comment]
@bwbarrett SPACC -- Suite of Point-to-point based Algorithms for Collective Communications

File: ompi/mca/coll/spacc/coll_spacc_allreduce.c [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+        int count_lhalf = count / 2;
+        int count_rhalf = count - count_lhalf;
+
+        if (rank % 2 != 0) {
+            /*
+             * Odd process -- exchange with rank - 1
+             * Send the left half of the input vector to the left neighbor,
+             * Recv the right half of the input vector from the left neighbor
+             */
+            err = MCA_PML_CALL(isend(rbuf, count_lhalf, dtype, rank - 1,
```
Comment (ggouaillardet) [inline ompi/mca/coll/spacc/coll_spacc_allreduce.c]: ```irecv() + send() + ompi_request_wait()``` is a bit faster that ```irecv() + isend() + ompi_request_wait_all(2, ...)```,
plus you do not have to worry about some error codes that are not supposed to be returned by collective.
last but not least, this is all automated in ```ompi_coll_base_sendrecv_actual()``` (or ```ompi_coll_base_sendrecv()``` if a MPI process might send/recv with itself

Comment (mkurnosov) [inline ompi/mca/coll/spacc/coll_spacc_allreduce.c]: Thanks @ggouaillardet, I modified implementation.

**Comment by bwbarrett** [comment]
@mkurnosov / @ggouaillardet, did you intend to merge this PR?  What are your next steps?

**Comment by ggouaillardet** [comment]
@mkurnosov could you please squash the two commits into one, then i will be happy to merge this into master.
@bwbarrett any objection against a PR vs the ```v3.x``` branch(es) ?

**Comment by mkurnosov** [comment]
@ggouaillardet I squashed commits.

**Comment by jsquyres** [comment]
This PR resulted in a bunch of new Coverity errors, including uninitialized variable warnings.

I don't know if this link will work for others, but this may show the spacc Coverity warnings: https://scan5.coverity.com/reports.htm#v22921/p10084

**Comment by mkurnosov** [comment]
@jsquyres The link does not work for me. I requested access to Open MPI on Coverity for viewing warnings.

**Comment by rhc54** [comment]
I approved your request

**Comment by jsquyres** [comment]
@mkurnosov Thanks!  Were you able to see the coverity errors inside that component?

**Comment by mkurnosov** [comment]
@jsquyres The reason of warnings is the possible negative result of the `opal_hibit(comm_size, comm->c_cube_dim + 1)`. But, such situation is not possible for this algorithms, comm_size is always > 1. 
For example, in the `mca_coll_basic_bcast_log_intra` such warning is eliminated by `assert(hibit >= 0)` before bitwise shift.

I prepared a branch with some fixes and now I am trying to recheck it on Coverity.

**Comment by ggouaillardet** [comment]
@mkurnosov iirc, one option is to add some appropriate ```assert()```, and they will silence coverity warnings

**Pull request merged**